### PR TITLE
Feature h3dbackbone

### PR DESCRIPTION
**Pull request by encore-zhou**

### CODE REVIEW

File: mmdet3d/models/backbones/multi_backbone.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+from mmdet.models import BACKBONES, build_backbone
+
+
+@BACKBONES.register_module()
+class MultiBackbone(nn.Module):
+    """MultiBackbone with different config. '''.
+
+    Args:
+        num_stream (int): The number of backbones.
```
Comment (xavierwu95) [inline mmdet3d/models/backbones/multi_backbone.py:15]: num_streams

File: mmdet3d/models/backbones/multi_backbone.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+        assert len(backbones) == num_stream
+
+        self.backbone_list = nn.ModuleList()
+        # Rename the ret_dict with different suffixs.
+        self.suffix_list = []
+
+        out_channels = 0
+
+        for bb_cfg in backbones:
```
Comment (xavierwu95) [inline mmdet3d/models/backbones/multi_backbone.py:49]: backbone_cfg

File: mmdet3d/models/backbones/multi_backbone.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        """Forward pass.
+
+        Args:
+            points (torch.Tensor): point coordinates with features,
+                with shape (B, N, 3 + input_feature_dim).
+
+        Returns:
+            dict[str, list[torch.Tensor]]: Outputs after SA and FP modules.
+
+                - fp_xyz (list[torch.Tensor]): The coordinates of \
```
Comment (xavierwu95) [inline mmdet3d/models/backbones/multi_backbone.py:101]: using  `\` to start a new line is not recommended

File: mmdet3d/models/backbones/multi_backbone.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    def forward(self, points):
+        """Forward pass.
+
+        Args:
+            points (torch.Tensor): point coordinates with features,
+                with shape (B, N, 3 + input_feature_dim).
+
+        Returns:
+            dict[str, list[torch.Tensor]]: Outputs after SA and FP modules.
+
```
Comment (xavierwu95) [inline mmdet3d/models/backbones/multi_backbone.py:100]: delete blank line

File: tests/test_backbones.py [hunk 5]
```diff
@@ -40,3 +40,103 @@ def test_pointnet2_sa_ssg():
...
+
+    # test list config
+    cfg_list = dict(
+        type='MultiBackbone',
+        num_stream=4,
+        backbones=[
+            dict(
+                type='PointNet2SASSG',
+                in_channels=4,
+                num_points=(2048, 1024, 512, 256),
```
Comment (xavierwu95) [inline tests/test_backbones.py:57]: Nit: reduce points or channels in unittest to increase the speed of ci.

File: tests/test_backbones.py [hunk 6]
```diff
@@ -40,3 +40,103 @@ def test_pointnet2_sa_ssg():
...
+    self.cuda()
+
+    assert len(self.backbone_list) == 4
+
+    xyz = np.fromfile('tests/data/sunrgbd/points/000001.bin', dtype=np.float32)
+    xyz = torch.from_numpy(xyz).view(1, -1, 6).cuda()  # (B, N, 6)
+    # test forward
+    ret_dict = self(xyz[:, :, :4])
+    hd_feature = ret_dict['hd_feature']
+    assert hd_feature.shape == torch.Size([1, 256, 1024])
```
Comment (xavierwu95) [inline tests/test_backbones.py:114]: also check the in_channels and out_channels of mlp layers

File: mmdet3d/models/backbones/multi_backbone.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    def forward(self, points):
+        """Forward pass.
+
+        Args:
+            points (torch.Tensor): point coordinates with features,
+                with shape (B, N, 3 + input_feature_dim).
+
+        Returns:
+            dict[str, list[torch.Tensor]]: Outputs after SA and FP modules.
+                - fp_xyz (list[torch.Tensor]): The coordinates of
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/multi_backbone.py:100]: Need an empty line between line 99 and 100

File: tests/test_backbones.py [hunk 8]
```diff
@@ -40,3 +40,107 @@ def test_pointnet2_sa_ssg():
...
+            radius=(0.2, 0.4, 0.8, 1.2),
+            num_samples=(64, 32, 16, 16),
+            sa_channels=((64, 64, 128), (128, 128, 256), (128, 128, 256),
+                         (128, 128, 256)),
+            fp_channels=((256, 256), (256, 256)),
+            norm_cfg=dict(type='BN2d'),
+            pool_mod='max',
+            suffix=''))
+
+    self = build_backbone(cfg_dict)
```
Comment (ZwwWayne) [inline tests/test_backbones.py:136]: self = build_backbone(cfg_dict).cuda()

File: mmdet3d/models/backbones/multi_backbone.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+from mmdet.models import BACKBONES, build_backbone
+
+
+@BACKBONES.register_module()
+class MultiBackbone(nn.Module):
+    """MultiBackbone with different config. '''.
+
+    Args:
+        num_streams (int): The number of backbones.
+        backbones (list or dict): A list of backbone configs.
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/multi_backbone.py:16]: We may add an additional arguments `suffixs = ['net0', 'net1']` thus the config used for backbones is just the same as the single backbone for a more clear usage.

File: mmdet3d/models/backbones/multi_backbone.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    """
+
+    def __init__(self,
+                 num_streams,
+                 backbones,
+                 aggregation_mlp_channels=None,
+                 conv_cfg=dict(type='Conv1d'),
+                 norm_cfg=dict(type='BN1d', eps=1e-5, momentum=0.01),
+                 act_cfg=dict(type='ReLU'),
+                 suffixes=['net0', 'net1'],
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/multi_backbone.py:33]: Use tuple, which is an immutable type and safer than `list`

File: mmdet3d/models/backbones/multi_backbone.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+from mmcv.cnn import ConvModule
+from mmcv.runner import load_checkpoint
+from torch import nn as nn
+
+from mmdet.models import BACKBONES, build_backbone
+
+
+@BACKBONES.register_module()
+class MultiBackbone(nn.Module):
+    """MultiBackbone with different config. '''.
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/multi_backbone.py:12]: Remove `'''`

File: mmdet3d/models/backbones/multi_backbone.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                 num_streams,
+                 backbones,
+                 aggregation_mlp_channels=None,
+                 conv_cfg=dict(type='Conv1d'),
+                 norm_cfg=dict(type='BN1d', eps=1e-5, momentum=0.01),
+                 act_cfg=dict(type='ReLU'),
+                 suffixes=('net0', 'net1'),
+                 **kwargs):
+        super().__init__()
+        assert isinstance(backbones, dict) or isinstance(backbones, list)
```
Comment (ZwwWayne) [inline mmdet3d/models/backbones/multi_backbone.py:36]: Test these assertions in line 36, 43, and 44 by passing wrong arguments with `with pytest.raises(AssertionError)`.

File: tests/test_backbones.py [hunk 13]
```diff
@@ -40,3 +40,116 @@ def test_pointnet2_sa_ssg():
...
+
+    assert len(self.backbone_list) == 2
+
+    # test forward
+    ret_dict = self(xyz[:, :, :4])
+
+    assert ret_dict['hd_feature'].shape == torch.Size([1, 128, 128])
+    assert ret_dict['fp_xyz_net0'][-1].shape == torch.Size([1, 128, 3])
+    assert ret_dict['fp_features_net0'][-1].shape == torch.Size([1, 256, 128])
+
```
Comment (ZwwWayne) [inline tests/test_backbones.py:144]: Add comments above on each `pytest.raises()` to tell what error you are testing.

**Review by ZwwWayne — APPROVED** [review]

**Comment by OpenMMLab-Assistant-007** [comment]
Hi!
@encore-zhou 
First of all, we want to express our gratitude for your significant PR in the OpenMMLab project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**